Dynamic api

mabruzzo opened the pull request:
This PR implements changes broadly described in Issue #100.

## Overview

This PR introduces some changes that make use of X-Macros (see the above issue for more details). This approach involves maintainence of a central list of all members of the ``chemistry_data`` struct (in ``src/clib/grackle_chemistry_data_fields.def``). By doing this we can:
  * drastically simplify the implementation of existing functions that access all the parameters in ``chemistry_data``
  * introduce an API for supporting dynamic access to members.

## Implementation Changes:

The ``_set_default_chemistry_parameters`` and ``show_parameters`` functions now make use of X-Macros and are now MUCH simpler.

I just wanted to suggest that it could make sense to delete ``set_default_chemistry_parameters.c`` and move the contents to ``initialize_chemistry_data.c``. Doing this would allow us to place the implementations of ``_set_default_chemistry_parameters`` and ``show_parameters`` next to each other, which might be useful for developers unfamiliar with this general technique. (Due to the simplifications to these functions -- the total length of ``initialize_chemistry_data.c`` after making this change would be shorter than it was before this PR).

I used [this script](https://pastebin.com/6rB5dZzD) as a sanity check to make sure this PR didn't alter the default values of any parameters.

I added a short section to the "How to Develop Grackle" page of the documentation called "Adding a New Parameter". Since the use of X-Macros is a little opaque, I wanted to make it clear to people that they should update the central list of fields.

## Dynamic API functions

I was primarily interested in adding the following 2 functions to dynamically facilitate access to ``int`` and ``double`` fields of a ``chemistry_data`` struct.
```c
int* local_chemistry_data_access_int(chemistry_data *my_chemistry,
                                     const char* param_name);
double* local_chemistry_data_access_double(chemistry_data *my_chemistry,
                                           const char* param_name);
```
The above functions return a ``NULL`` pointer if ``my_chemistry`` is ``NULL``, ``param_name`` isn't a known parameter, or the function for the wrong type is used to access ``param_name`` (e.g. if ``local_chemistry_data_access_int`` is used to access ``Gamma``).

While trying to test the above functions, I realized it would be useful to have functions for querying the names of parameters. The following functions are used to query the name of the ith parameter of a given type.
```c
const char* param_name_int(size_t i);
const char* param_name_double(size_t i);
```
The order of the parameters directly depends on the order that they are listed in ``src/clib/grackle_chemistry_data_fields.def``. Therefore, I have added notes in the documentation stating that the exact order may not be stable between different versions of grackle. As an aside, the inclusion of these latter functions will make (de)serialization of ``chemistry_data`` much easier (which was something that was very tedious in Enzo-E).

I added a section to the "Adding Grackle to Your Simulation Code" page of the documentation that talks about how to use the new functions that provide dynamic access to the fields ``chemistry_data`` and why it can be useful.

Let me know if you want me to change the names of these functions or change their signature in some way. To test these functions, I have two helper functions to pygrackle and added ``src/python/tests/test_dynamic_api.py``. Let me know if you are unhappy with this testing approach and I can change it.

I have included some additional information about the new functions in the following spoiler tag, which may or may not be useful for reviewing this PR.

<details>
  <summary>Missing analogs for string parameters</summary>

  A more complete API would also include the following two functions:
  ```c
  char** local_chemistry_data_access_string(chemistry_data *my_chemistry,
                                            const char* param_name);
  const char* param_name_string(size_t i);
  ```

  Because ``chemistry_data`` just has one string parameter, ``grackle_data_file``, I decided to omit these to reduce the number of new functions. With that said, I'm happy to add these functions.
</details>

<details>
  <summary>Speed of the functions & alternative </summary>

  The current implementations of ``param_name_int`` and ``param_name_double``, are pretty dumb. The computational complexity is ``O(i)``.

  An alternative implementation of the ``param_name_*`` functions with ``O(1)`` complexity is illustrated [here](https://pastebin.com/eYY5Y2m1) (I haven't tried compiling it - so it may need debugging). This alternative also speeds up ``local_chemistry_data_access_``. The tradeoff is another usage of X-Macros. If you prefer this alternative, let me know and I'll swap it in.

  The functions for accessing fields has ``O(N)`` complexity, where ``N`` is the number of fields in ``chemistry_data``. This is probably fine, since ``chemistry_data`` usually is just configured once. However, in the future, we could make them have ``O(log(N))`` complexity if we ensured the entries in ``src/clib/grackle_chemistry_data_fields.def`` were alphabetized (this would be easy to check in a test case).
</details>

<details>
  <summary>Using dynamic API in Pygrackle?</summary>

  It would be straight-forward to alter the internals of the ``chemistry_data`` extension type to make use of this new dynamic API. This could be done by defining custom ``__getattr__`` and ``__setattr__`` functions.

  The major benefit to doing that it removes all of the necessary maintenance required to make newly added fields of ``chemistry_data`` accessible through the ``pygrackle`` interface (i.e. it would no longer be necessary to declare every field in the ``chemistry_data`` struct in ``grackle_defs.pxd`` and to add properties for each field in the ``chemistry_data`` extension type in ``grackle_wrapper.pyx``). We could probably also ditch most/all of the new testing code added in this PR since all the other python tests would implicitly test the dynamic API.

  Of course, the tradeoff is that access to the fields of the ``chemistry_data`` struct would now be slightly slower. 
</details>

brittonsmith commented:
@mabruzzo, this is really nice. I'm working on a review, but it's taking a while with everything else going on here. In the meantime, I'll comment on the additional points you left at the bottom of the description in the case that you want to do anything more with them while I finish the rest of the review.

If you're up for it, I would be in favor of adding an analogous function for string parameters for overall consistency. I can't think of many additional strings we might want to carry, but perhaps additional data files? Either way, getting and setting strings has enough complication that a robust API would be a solid improvement.

I don't have a strong opinion on optimizing the functions as I don't think it will ever be a considerable portion of the cost for a computation involving grackle. That said, the quicker implementation did not look too complicated, so if you like it better, I'm happy if you want to go with that instead.

Finally, I'd love to see the dynamic api used in pygrackle if it could cut down on all the boilerplate for adding new parameters. It's a difficult part of the code to work in. The less anyone has to do it, the better.

Anyway, this is a great enhancement. I'd like to see it get in before the next release. I'll get the rest of the review done soon.

mabruzzo commented:
I added the analogous functions for accessing strings and took a stab at refactoring the python interface to make use of the dynamic api.

Refactoring the python interface was more complicated than I had expected. The complications stem from difficulties with introducing a custom `__setattr__(self, name, value)` method to an extension type, that has special behavior a subset of attribute names and then falls back to the default behavior of `__setattr__` (which is necessary to access properties).
- In an ordinary python class, you can reproduce the default behavior by simply calling `super().__setattr__(self, name, value)` within `__setattr__`.
- However, this does not work in an extension class. I ultimately achieved the objective by manually reimplementing the default behavior of `__setattr__`.

(As an aside, there aren't analogous problems with modifying `__getattr__`/`__getattribute__` - there's a much easier workaround).

I am not thrilled with my custom implementation of `__setattr__`. I tried to be as clear as possible about why we were doing certain things in comments. But, if you would prefer not to go this route, I totally understand. I sketched out an alternative solution to this problem, [here](https://pastebin.com/NwPtvJLU), by introducing an extra layer of indirection. Let me know your preference (note that the alternative implementation may still have typos/bugs)

If I have time, I plan to optimize the dynamic api functions. But, I am going to be busy this week and may not have time. So please don't wait for me to do that before reviewing this PR (since it won't affect the API, I can always do it later)

mabruzzo commented:
Closing this PR, because I seemed to have introduced some kind of problem into the commit history of this branch.  This issue causes both the circleci and readthedocs checks to fail with an error that looked like the following:
> `fatal: reference is not a tree: c4860706785473f750c77317c9f9a54622f33eb4`

This issue arose following a commit in which I merged changes from the main branch. I believe I caused this issue because I had accidentally performed a `git stash` while resolving the merge conflicts (the issue might be more related to what I did next -- I think after popping the stashed changes I reset the changed files and initiated a new merge).

To try to resolve this issue, I did the following:

- First, I tried resetting the commit history to the commit just before the problematic merge and executed a new merge. Then I force pushed the changes. This didn't work
- Next, I tried creating a fresh branch and introduced all the changed files in a brand new single commit. Force pushing that new branch didn't solve the problem either. This led me to conclude that I may have also introduced some kind of error into GitHub's machinery for the PR itself.

Because neither of the above actions worked, I choose to create a fresh PR from the fresh branch. Please see PR #130.
